modules/mobile-agents/android/client/plugins/IDPProxy/src/org/wso2/mdm/agent/proxy/APIController.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2014, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ * 
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.mdm.agent.proxy;
+
+import android.content.Context;
+import android.os.AsyncTask;
+import android.util.Log;
+import org.wso2.mdm.agent.proxy.beans.EndPointInfo;
+import org.wso2.mdm.agent.proxy.beans.Token;
+import org.wso2.mdm.agent.proxy.interfaces.APIResultCallBack;
+import org.wso2.mdm.agent.proxy.interfaces.TokenCallBack;
+import org.wso2.mdm.agent.proxy.utils.ServerUtilities;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class APIController implements TokenCallBack {
+	private static final String TAG = "APIController";
+	private static Token token;
+	private static String clientKey, clientSecret;
+	private APIResultCallBack apiResultCall;
+	private EndPointInfo apiUtilitiesCurrent;
+
+	public void setClientDetails(String clientKey, String clientSecret) {
+		APIController.clientKey = clientKey;
+		APIController.clientSecret = clientSecret;
+	}
+
+	/**
+	 * Invoking an API using retrieved token.
+	 *
+	 * @param apiUtilities      - Server and API end point information.
+	 * @param apiResultCallBack - API result callback data.
+	 * @param requestCode       - Request code to avoid response complications.
+	 * @param context           - Application context.
+	 */
+	public void invokeAPI(EndPointInfo apiUtilities, APIResultCallBack apiResultCallBack,
+	                      int requestCode, Context context) {
+
+		apiResultCall = apiResultCallBack;
```

Comment:
Do we need to globally maintain the state of apiResultCall? In a typical concurrent environment, this would most certainly leave the application in an inconsistent state due to possible concurrency issues.